src/System.Private.CoreLib/shared/System/Number.Grisu3.cs

```diff
@@ -311,410 +308,621 @@ internal static unsafe class Grisu3
+            //
+            // The rounding might shift the whole buffer, in which case, the kappy is adjusted.
+            // For example "99", kappa = 3 might become "10", kappa = 4.
+            //
+            // If (2 * rest) > tenKappa then the buffer needs to be round up.
+            // rest can have an error of +/- 1 unit.
+            // This function accounts for the imprecision and returns false if the rounding direction cannot be unambiguously determined.
+            //
+            // Preconditions:
+            //      rest < tenKappa
+            private static bool RoundWeedCounted(Span<byte> buffer, int length, ulong rest, ulong tenKappa, ulong unit, ref int kappa)
             {
                 Debug.Assert(rest < tenKappa);
 
-                // 1. tenKappa <= ulp: we don't have an idea which way to round.
-                // 2. Even if tenKappa > ulp, but if tenKappa <= 2 * ulp we cannot find the way to round.
-                // Note: to prevent overflow, we need to use tenKappa - ulp <= ulp.
-                if ((tenKappa <= ulp) || ((tenKappa - ulp) <= ulp))
+                // The following tests are done in a specific order to avoid overflows.
+                // They will work correctly with any ulong values of rest < tenKappa and unit.
+                //
+                // If the unit is too big, then we don't know which way to round.
+                // For example, a unit of 50 means that the real number lies within rest +/- 50.
+                // If 10^kappa == 40, then there is no way to tell which way to round.
+                //
+                // Even if unit is just half the size of 10^kappa we are already completely lost.
+                // And after the previous test, we know that the expression will not over/underflow.
+                if ((unit >= tenKappa) || ((tenKappa - unit) <= unit))
                 {
                     return false;
                 }
 
-                // tenKappa >= 2 * (rest + ulp). We should round down.
-                // Note: to prevent overflow, we need to check if tenKappa > 2 * rest as a prerequisite.
-                if (((tenKappa - rest) > rest) && ((tenKappa - (2 * rest)) >= (2 * ulp)))
+                // If 2 * (rest + unit) <= 10^kappa, we can safely round down.
+                if (((tenKappa - rest) > rest) && ((tenKappa - (2 * rest)) >= (2 * unit)))
                 {
                     return true;
                 }
 
-                // tenKappa <= 2 * (rest - ulp). We should round up.
-                // Note: to prevent overflow, we need to check if rest > ulp as a prerequisite.
-                if ((rest > ulp) && (tenKappa <= (rest - ulp) || ((tenKappa - (rest - ulp)) <= (rest - ulp))))
+                // If 2 * (rest - unit) >= 10^kappa, we can safely round up.
+                if ((rest > unit) && (tenKappa <= (rest - unit) || ((tenKappa - (rest - unit)) <= (rest - unit))))
```

Comment:
cache `rest - unit` in a local

Reply:
This is a place where the JIT looks to do the optimization already. It would be nice to keep our port inline with the reference code where possible.